### PR TITLE
prov/efa: adjust multi req data size

### PR DESCRIPTION
**Pull request by wzamazon**
adjust the data size of multiple req sent via one call to ibv_post_send.

### CODE REVIEW

File: prov/efa/src/rxr/rxr_ep.c [hunk 1]
```diff
@@ -389,6 +389,7 @@ void rxr_tx_entry_init(struct rxr_ep *ep, struct rxr_tx_entry *tx_entry,
 		tx_op_flags &= ~FI_COMPLETION;
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_ep.c:389]: this commit message has a typo: effcient -> efficient

Also I don't understand what:
"the size of these packet to be close" means, can you elaborate? Reading the code it looks like you mean something like, keep data packet sizes as similar as possible.

File: prov/efa/src/rxr/rxr_op_entry.c [hunk 2]
```diff
@@ -216,3 +216,43 @@ size_t rxr_tx_entry_max_req_data_capacity(struct rxr_ep *ep, struct rxr_op_entry
 						      header_flags,
 						      tx_entry->rma_iov_count);
 }
+
+/**
+ * @breif set the max_req_data_size field of a tx_entry for multi-req
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_op_entry.c:221]: brief*

File: prov/efa/src/rxr/rxr_op_entry.c [hunk 3]
```diff
@@ -216,3 +216,43 @@ size_t rxr_tx_entry_max_req_data_capacity(struct rxr_ep *ep, struct rxr_op_entry
...
+
+	assert(rxr_pkt_is_mulreq(pkt_type));
+
+	max_req_data_capacity = rxr_tx_entry_max_req_data_capacity(ep, tx_entry, pkt_type);
+	assert(max_req_data_capacity);
+
+	mulreq_total_data_size = rxr_op_entry_mulreq_total_data_size(tx_entry, pkt_type);
+	assert(mulreq_total_data_size);
+
+	num_req = (mulreq_total_data_size - 1)/max_req_data_capacity + 1;
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_op_entry.c:253]: why is this calculation different from rxr_tx_entry_num_req. Shouldn't you just use that function here?

**Review by wckzhang — CHANGES_REQUESTED** [review]

File: prov/efa/src/rxr/rxr_ep.c (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:389]: Yes. I want to make the data size as similar as possible.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:389]: Updated the comment and commit title.

File: prov/efa/src/rxr/rxr_op_entry.c (hunk 2 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_op_entry.c:221]: Thank you! Fixed.

File: prov/efa/src/rxr/rxr_op_entry.c [hunk 4]
```diff
@@ -216,3 +216,43 @@ size_t rxr_tx_entry_max_req_data_capacity(struct rxr_ep *ep, struct rxr_op_entry
...
+
+	tx_entry->max_req_data_size = ofi_get_aligned_size((mulreq_total_data_size - 1)/num_req + 1, MEMORY_ALIGNMENT);
+	if (tx_entry->max_req_data_size >  max_req_data_capacity)
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_op_entry.c:256]: Thank you! fixed.

File: prov/efa/src/rxr/rxr_op_entry.c (hunk 3 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_op_entry.c:253]: No. That function uses `rxr_tx_entry->max_req_data_size`, which is set by this function.

At the very begining, we do not know how many packets we are  going to send. so we need to determine that by using data_capacity.

Once we know the number of packet, we can calculate the average packet size then round up to alignment. 

we then calculate the number of packet again, which is `rxr_op_entry_num_req` for

**Review by wckzhang — APPROVED** [review]

**Comment by wzamazon** [comment]
Will add a unit test as a follow up PR.

**Pull request merged**